common/src/Common/Wallet.hs

```diff
@@ -384,13 +386,34 @@ instance FromJSON key => FromJSON (Key key) where
       { _key_pair = pair
       }
 
-
 type KeyStorage key = IntMap (Key key)
 
+-- | Predefined predicates in Pact.
+--
+--   Userdefined ones are possible too, although the UI currently does not support them.
+predefinedPreds :: [ Text ]
+predefinedPreds = [ "keys-all", "keys-2", "keys-any" ]
+
 filterKeyPairs :: Set PublicKey -> IntMap (Key key) -> [KeyPair key]
 filterKeyPairs s m = Map.elems $ Map.restrictKeys (toMap m) s
   where toMap = Map.fromList . fmap (\k -> (_keyPair_publicKey $ _key_pair k, _key_pair k)) . IntMap.elems
 
+keysetSatisfiesPredicate :: AddressKeyset -> IntMap (Key key) -> Bool
+keysetSatisfiesPredicate ak keys0
+  | addpred <- _addressKeyset_pred ak, addpred `elem` predefinedPreds =
+      let
+        keys = Set.fromList $ fmap (_keyPair_publicKey . _key_pair) $ IntMap.elems keys0
+        kskeys = _addressKeyset_keys ak
+        keyIntersections = Set.intersection kskeys keys
+      in
+        case _addressKeyset_pred ak of
+          "keys-all" -> kskeys == keyIntersections
+          "keys-any" -> not $ Set.null keyIntersections
+          "keys-2" -> length keyIntersections >= 2
+          _ -> False
+  | otherwise =
+      not $ null $ filterKeyPairs (_addressKeyset_keys ak) keys0
```

Comment:
Shouldn't this be the `_ ->` inside the case? The guard doesn't make much sense to me